File: skills/inspect-and-debug/harness-doctor/scripts/lib/render.ts
/**
 * Renders the aggregate report as compact markdown, for a model to read.
 *
 * Dense tables, numbers with units, no decoration. Every table states how many rows
 * it dropped, so a truncated ranking can never read as a complete one. The whole
 * summary stays near 400 lines: rankings are capped again at render time.
 */

import { cell, int, pct, plural, sec, table, tok, usd } from "./format.ts";
import { cmpStr } from "./stats.ts";
import type { AggregateReport, Ranking, SessionKind, TokenTotals } from "./types.ts";

/** How the summary is rendered. */
export interface RenderOptions {
  /** Rows per table. Clamped to 25: the summary is a briefing, not a data dump. */
  top?: number;
  /** Printed as the last line when set, so a caller can find the fact tables. */
  workDir?: string;
}

/** Hard cap on rows per table, whatever `--top` says. */
const MAX_ROWS = 25;

/** Rows to render from a ranking, after the render-time cap. */
const take = <T>(r: Ranking<T>, limit: number) => r.rows.slice(0, limit);

/**
 * One line saying what a table left out, counting both the aggregate cap and the
 * render cap. Returns an empty list when every row is on screen.
 */
const dropped = <T>(r: Ranking<T>, limit: number) => {
  const shown = Math.min(r.rows.length, limit);
  const hidden = r.total - shown;
  return hidden > 0 ? [`${shown} of ${int(r.total)} rows, ${int(hidden)} not shown.`] : [];
};

/** A ranked table plus its dropped-rows line. */
const ranked = <T>(
  r: Ranking<T>,
  limit: number,
  headers: readonly string[],
  row: (x: T) => (string | number)[],
) => [...table(headers, take(r, limit).map(row)), ...dropped(r, limit)];

/** Token totals as a one-line "in / out / read / write" string. */
const totalsLine = (t: TokenTotals) =>
  `in ${tok(t.in)}, out ${tok(t.out)}, cacheRead ${tok(t.cacheRead)}, cacheWrite ${tok(t.cacheCreate)}`;

/** Sum of every token class in a totals block. */
const allTokens = (t: TokenTotals) => t.in + t.out + t.cacheRead + t.cacheCreate;

/** Session kinds in fixed order. */
const KINDS: readonly SessionKind[] = ["main", "subagent", "workflow-agent"];

/** Record entries sorted by value descending, then key, capped. */
const byValue = (rec: Record<string, number>, limit: number) =>
  Object.entries(rec)
    .sort((a, b) => b[1] - a[1] || cmpStr(a[0], b[0]))
    .slice(0, limit);

/** Window and the numbers worth reading first. */
const headline = (r: AggregateReport) => {
  const w = r.window;
  const kinds = KINDS.map((k) => `${k} ${int(w.sessionsByKind[k] ?? 0)}`).join(", ");
  const flail = r.commands.flagFlailing;
  const retry = r.errors.retryLoops;
  const corrections = r.human.corrections;
  const correctionCount = corrections.rows.reduce((a, b) => a + b.count, 0);
  return [
    "# harness audit",
    "",
    `Window ${w.days}d, ${w.fromIso || "unknown"} to ${w.toIso || "unknown"}. Files scanned ${int(w.filesScanned)}, skipped ${int(w.filesSkipped)}. Sessions: ${kinds}. Projects ${int(w.projectCount)}.`,
    "",
    "## headline",
    "",
    ...table(
      ["metric", "value"],
      [
        ["tokens total", tok(allTokens(r.tokens.totals))],
        ["estimated cost", usd(r.tokens.estCostUsd)],
        ["tool wall-clock (approx)", sec(r.wallClock.totalToolSec)],
        [
          "bash seconds",
          sec(r.wallClock.byTool.rows.find((t) => t.tool === "Bash")?.totalSec ?? 0),
        ],
        [
          "commands re-run in-session",
          `${plural(r.commands.repeatsInSession.rows.reduce((a, b) => a + b.occurrences, 0), "repeat")} over ${plural(r.commands.repeatsInSession.total, "command")}`,
        ],
        [
          "flag flailing",
          `${plural(flail.rows.reduce((a, b) => a + b.occurrences, 0), "distinct command")} across ${plural(flail.total, "bin")}`,
        ],
        [
          "retry loops",
          `${plural(retry.rows.reduce((a, b) => a + b.count, 0), "loop")} over ${plural(retry.total, "family", "families")}`,
        ],
        ["web repeat rate", `${pct(r.web.repeatRate)} of ${int(r.web.totalCalls)} calls`],
        [
          "fan-out prefix rewritten",
          `${tok(r.fanout.totalFirstTurnCacheCreate)} cacheWrite on first turns, ${int(r.fanout.agentsWithFirstTurnCacheWrite)} agents`,
        ],
        [
          "human corrections",
          `${int(correctionCount)} bucket hits over ${int(r.human.humanTurnCount)} human turns`,
        ],
        ["interrupts", int(r.human.interrupts)],
      ],
    ),
    "",
  ];
};

/** Tokens by kind, model, and project, with the cost estimate. */
const tokensBlock = (r: AggregateReport, limit: number) => {
  const t = r.tokens;
  const unpriced =
    t.unpricedModels.length > 0
      ? [`Unpriced model ids, charged at sonnet rates: ${t.unpricedModels.join(", ")}.`]
      : ["Every model id matched a priced family."];
  return [
    "## tokens",
    "",
    `Totals: ${totalsLine(t.totals)}. Estimated cost ${usd(t.estCostUsd)}.`,
    "",
    ...table(
      ["kind", "in", "out", "cacheRead", "cacheWrite"],
      KINDS.map((k) => {
        const v = t.byKind[k] ?? { in: 0, out: 0, cacheRead: 0, cacheCreate: 0 };
        return [k, tok(v.in), tok(v.out), tok(v.cacheRead), tok(v.cacheCreate)];
      }),
    ),
    "",
    "By model, tokens attributed in proportion to model calls per session:",
    "",
    ...table(
      ["model", "in", "out", "cacheRead", "cacheWrite"],
      Object.entries(t.byModel)
        .sort((a, b) => allTokens(b[1]) - allTokens(a[1]) || cmpStr(a[0], b[0]))
        .slice(0, limit)
        .map(([model, v]) => [
          model,
          tok(v.in),
          tok(v.out),
          tok(v.cacheRead),
          tok(v.cacheCreate),
        ]),
    ),
    "",
    ...ranked(
      t.byProject,
      limit,
      ["project", "sessions", "tokens", "cacheRead", "est cost"],
      (p) => [
        p.proj,
        int(p.sessions),
        tok(allTokens(p.totals)),
        tok(p.totals.cacheRead),
        usd(p.estCostUsd),
      ],
    ),
    "",
    ...unpriced,
    t.note,
    "",
  ];
};

/** Wall-clock by tool and project. */
const wallClockBlock = (r: AggregateReport, limit: number) => [
  "## wall clock",
  "",
  `Total tool time ${sec(r.wallClock.totalToolSec)}. ${r.wallClock.note}`,
  "",
  ...ranked(
    r.wallClock.byTool,
    limit,
    ["tool", "calls", "total", "median", "p95", "err"],
    (x) => [x.tool, int(x.calls), sec(x.totalSec), sec(x.medianSec), sec(x.p95Sec), pct(x.errRate)],
  ),
  "",
  ...ranked(r.wallClock.byProject, limit, ["project", "calls", "total"], (x) => [
    x.proj,
    int(x.calls),
    sec(x.totalSec),
  ]),
  "",
];

/** Command families, repeats, and flag flailing. */
const commandsBlock = (r: AggregateReport, limit: number) => [
  "## commands",
  "",
  ...ranked(
    r.commands.byTotalSec,
    limit,
    ["family", "bin", "calls", "total", "median", "p95", "err"],
    (x) => [
      x.family,
      x.bin,
      int(x.count),
      sec(x.totalSec),
      sec(x.medianSec),
      sec(x.p95Sec),
      pct(x.errRate),
    ],
  ),
  "",
  "Most-run families:",
  "",
  ...ranked(r.commands.byCount, limit, ["family", "calls", "total", "err"], (x) => [
    x.family,
    int(x.count),
    sec(x.totalSec),
    pct(x.errRate),
  ]),
  "",
  "Identical command re-run inside one session, counting every run after the first:",
  "",
  ...ranked(
    r.commands.repeatsInSession,
    limit,
    ["command", "repeats", "sessions", "wasted"],
    (x) => [x.cmd, int(x.occurrences), int(x.sessions), sec(x.wastedSec)],
  ),
  "",
  "Flag flailing: same binary, 4 or more distinct normalised commands inside 10 minutes of one session.",
  "",
  ...ranked(
    r.commands.flagFlailing,
    limit,
    ["bin", "distinct cmds", "sessions", "total", "sample"],
    (x) => [
      x.bin,
      int(x.occurrences),
      int(x.sessions),
      sec(x.totalSec),
      x.sampleCmds.slice(0, 2).join(" ; "),
    ],
  ),
  "",
];

/** Error clusters, missing binaries, refusals, and retry loops. */
const errorsBlock = (r: AggregateReport, limit: number) => [
  "## errors",
  "",
  ...ranked(
    r.errors.clusters,
    limit,
    ["signature", "count", "sessions", "top tool", "sample"],
    (x) => [x.sig, int(x.count), int(x.sessionCount), x.topTool, x.sampleMsg],
  ),
  "",
  "Binaries the shell could not find:",
  "",
  ...ranked(r.errors.missingBinaries, limit, ["bin", "count", "sessions", "sample"], (x) => [
    x.bin,
    int(x.count),
    int(x.sessions),
    x.sampleMsg,
  ]),
  "",
  "Refused calls:",
  "",
  ...ranked(r.errors.permissionDenied, limit, ["target", "count", "sessions", "sample"], (x) => [
    x.target,
    int(x.count),
    int(x.sessions),
    x.sampleMsg,
  ]),
  "",
  "Retry loops: a command re-run within 5 minutes of failing.",
  "",
  ...ranked(
    r.errors.retryLoops,
    limit,
    ["family", "loops", "median retries", "sessions", "sample"],
    (x) => [x.family, int(x.count), x.medianRetries.toFixed(1), int(x.sessions), x.sampleCmd],
  ),
  "",
];

/** Context pressure: peak sessions, big outputs, re-reads, cache write ratio. */
const contextBlock = (r: AggregateReport, limit: number) => {
  const shown = take(r.context.topSessions, limit);
  const shownRead = shown.reduce((a, s) => a + s.cacheReadTokens, 0);
  const share = r.tokens.totals.cacheRead > 0 ? shownRead / r.tokens.totals.cacheRead : 0;
  return [
    "## context",
    "",
    `The ${shown.length} sessions with the largest peak context hold ${pct(share)} of all cache-read tokens (${tok(shownRead)} of ${tok(r.tokens.totals.cacheRead)}).`,
    "",
    ...ranked(
      r.context.topSessions,
      limit,
      ["session", "project", "kind", "maxCtx", "cacheRead", "est cost"],
      (x) => [
        x.sid.slice(0, 8),
        x.proj,
        x.kind,
        tok(x.maxCtx),
        tok(x.cacheReadTokens),
        usd(x.estCostUsd),
      ],
    ),
    "",
    "Single tool results that returned the most text:",
    "",
    ...ranked(r.context.bigOutputs, limit, ["tool", "arg", "chars", "tokens est"], (x) => [
      x.tool,
      x.arg ?? "",
      int(x.outChars),
      tok(x.tokensEst),
    ]),
    "",
    ...table(
      ["tool", "p95 chars"],
      byValue(r.context.p95CharsByTool, limit).map(([tool, v]) => [tool, int(v)]),
    ),
    "",
    "Same file read more than once in one session:",
    "",
    ...ranked(r.context.reReads, limit, ["file", "extra reads", "sessions", "wasted tokens"], (x) => [
      x.arg,
      int(x.occurrences),
      int(x.sessions),
      tok(x.wastedTokensEst),
    ]),
    "",
    ...table(
      ["kind", "cacheWrite share"],
      KINDS.map((k) => [k, pct(r.context.cacheWriteRatio[k] ?? 0)]),
    ),
    "",
  ];
};

/** Web traffic, repeats, and concentration by run. */
const webBlock = (r: AggregateReport, limit: number) => [
  "## web",
  "",
  `${int(r.web.totalCalls)} calls, ${sec(r.web.totalSec)}, ${tok(r.web.totalChars / 4)} tokens of text. ${pct(r.web.repeatRate)} of calls re-fetched a url or query already answered.`,
  "",
  ...table(
    ["tool", "calls", "sec", "chars"],
    Object.entries(r.web.byTool).map(([tool, s]) => [
      tool,
      int(s.calls),
      sec(s.sec),
      int(s.chars),
    ]),
  ),
  "",
  ...ranked(r.web.repeats, limit, ["tool", "url or query", "fetches", "chars", "one run"], (x) => [
    x.tool,
    x.arg,
    int(x.count),
    int(x.chars),
    x.sameRun ? "yes" : "no",
  ]),
  "",
  ...ranked(r.web.byRun, limit, ["run", "calls", "agents", "chars"], (x) => [
    x.run,
    int(x.calls),
    int(x.agents),
    int(x.chars),
  ]),
  "",
];

/** Fan-out cost per run, and agents that read a lot and wrote little. */
const fanoutBlock = (r: AggregateReport, limit: number) => [
  "## fanout",
  "",
  `${int(r.fanout.agentsWithFirstTurnCacheWrite)} agents paid for a fresh prefix on their first turn, ${tok(r.fanout.totalFirstTurnCacheCreate)} cacheWrite tokens in total.`,
  "",
  ...ranked(
    r.fanout.runs,
    limit,
    ["run", "project", "agents", "tokens", "est cost", "median first-turn write"],
    (x) => [
      x.run,
      x.proj,
      int(x.agents),
      tok(allTokens(x.totals)),
      usd(x.estCostUsd),
      tok(x.medianFirstTurnCacheCreate),
    ],
  ),
  "",
  "Low-yield agents: bottom decile of output tokens per cache-read token.",
  "",
  ...ranked(
    r.fanout.lowYield,
    limit,
    ["agent", "run", "cacheRead", "out", "yield", "est cost"],
    (x) => [
      x.sid.slice(0, 8),
      x.run ?? "",
      tok(x.cacheReadTokens),
      tok(x.outTokens),
      x.yield.toFixed(4),
      usd(x.estCostUsd),
    ],
  ),
  "",
];

/** Corrections by theme, with verbatim samples, and requests made twice. */
const humanBlock = (r: AggregateReport, limit: number) => {
  const buckets = take(r.human.corrections, limit).flatMap((b) => [
    `- ${b.bucket}: ${plural(b.count, "turn")}`,
    ...b.samples.map((s) => `  - "${cell(s, 200)}"`),
  ]);
  return [
    "## human",
    "",
    `${int(r.human.humanTurnCount)} human turns, ${int(r.human.interrupts)} interrupts. Only turns flagged as real human input are scanned: harness echoes, fenced code, and quoted lines are excluded.`,
    "",
    "Corrections by theme, a turn may match more than one theme:",
    "",
    ...(buckets.length > 0 ? buckets : ["(none)"]),
    ...dropped(r.human.corrections, limit),
    "",
    "Near-duplicate requests the human made in more than one session:",
    "",
    ...ranked(r.human.repeatedRequests, limit, ["request", "turns", "sessions"], (x) => [
      cell(x.label, 90),
      int(x.count),
      int(x.sessions),
    ]),
    "",
  ];
};

/** Per project: cost, bash time, top commands, and the rules files on disk. */
const projectsBlock = (r: AggregateReport, limit: number) => [
  "## projects",
  "",
  ...ranked(
    r.projects,
    limit,
    ["project", "sessions", "tokens", "est cost", "bash", "top commands", "on disk", "pkg"],
    (p) => [
      p.proj,
      int(p.sessions),
      tok(allTokens(p.totals)),
      usd(p.estCostUsd),
      sec(p.bashSec),
      p.topCommands
        .slice(0, 3)
        .map((c) => `${c.family} x${c.count}`)
        .join(" ; "),
      [
        p.onDisk.claudeMd ? "CLAUDE.md" : "",
        p.onDisk.agentsMd ? "AGENTS.md" : "",
        p.onDisk.settingsJson ? "settings.json" : "",
      ]
        .filter(Boolean)
        .join(" ") || "none",
      `${p.onDisk.packageManager ?? "-"}: ${p.onDisk.scriptNames.slice(0, 6).join(",") || "-"}`,
    ],
  ),
  "",
];

/**
 * Render the whole report as markdown. Row caps apply again here, and every table
 * says how many rows it dropped, so nothing reads as complete when it is not.
 */
export const renderMarkdown = (report: AggregateReport, options: RenderOptions = {}) => {
  const limit = Math.max(1, Math.min(MAX_ROWS, Math.floor(options.top ?? 15)));
  const lines = [
    ...headline(report),
    ...tokensBlock(report, limit),
    ...wallClockBlock(report, limit),
    ...commandsBlock(report, limit),
    ...errorsBlock(report, limit),
    ...contextBlock(report, limit),
    ...webBlock(report, limit),
    ...fanoutBlock(report, limit),
    ...humanBlock(report, limit),
    ...projectsBlock(report, limit),
  ];
  if (options.workDir) lines.push(options.workDir);
  return `${lines.join("\n").replace(/\n{3,}/g, "\n\n").trimEnd()}\n`;
};

File: skills/inspect-and-debug/harness-doctor/scripts/lib/render.py
"""Render the aggregate dict as JSON or compact markdown."""

import json


def render_json(agg) -> str:
    """schemaVersion:1 wrapper plus the aggregate sections, pretty-printed."""
    out = {"schemaVersion": 1, **agg}
    return json.dumps(out, indent=2, sort_keys=False)


def _n(v):
    if isinstance(v, float):
        return f"{v:,.3f}"
    return f"{v:,}"


def _usd(v):
    return f"${v:,.4f}"


def _table(headers, rows):
    if not rows:
        return "_none_\n"
    lines = ["| " + " | ".join(headers) + " |", "|" + "|".join(["---"] * len(headers)) + "|"]
    for row in rows:
        cells = [str(c) if c is not None else "" for c in row]
        lines.append("| " + " | ".join(cells) + " |")
    return "\n".join(lines) + "\n"


def _dropped_note(dropped):
    return f"\n_{dropped} more row(s) not shown (see fact tables)._\n" if dropped else ""


def _render_window(w):
    return (
        f"## Window\n\n"
        f"Days: {w['days']}. From {w['fromIso']} to {w['toIso']}.\n"
        f"Files scanned: {_n(w['filesScanned'])}. Files skipped: {_n(w['filesSkipped'])}. "
        f"Projects: {_n(w['projectCount'])}.\n"
        f"Sessions by kind: main {_n(w['sessionsByKind']['main'])}, "
        f"subagent {_n(w['sessionsByKind']['subagent'])}, "
        f"workflow-agent {_n(w['sessionsByKind']['workflow-agent'])}.\n"
    )


def _render_tokens(t, top_n):
    out = [f"## Tokens (est. cost is an estimate, top {top_n} shown per ranking)\n"]
    tot = t["totals"]
    out.append(
        f"Totals: in {_n(tot['in'])}, out {_n(tot['out'])}, cacheRead {_n(tot['cacheRead'])}, "
        f"cacheCreate {_n(tot['cacheCreate'])}. Estimated cost: {_usd(t['estCostUsd'])}.\n"
    )
    if t["unpricedModels"]:
        out.append(f"Unpriced models (priced as sonnet, check rates): {', '.join(t['unpricedModels'])}.\n")
    out.append("\nBy model:\n")
    rows = [(m, _n(v["in"]), _n(v["out"]), _n(v["cacheRead"]), _n(v["cacheCreate"]), _usd(v["estCostUsd"]))
            for m, v in t["byModel"].items()]
    out.append(_table(["model", "in", "out", "cacheRead", "cacheCreate", "estCostUsd"], rows))
    out.append("\nBy kind:\n")
    rows = [(k, _n(v["in"]), _n(v["out"]), _n(v["cacheRead"]), _n(v["cacheCreate"])) for k, v in t["byKind"].items()]
    out.append(_table(["kind", "in", "out", "cacheRead", "cacheCreate"], rows))
    out.append(f"\nTop projects by total tokens ({top_n} of {len(t['byProject']) + t['byProjectDroppedCount']}):\n")
    rows = [(p["proj"], _n(p["totalTokens"]), _n(p["cacheRead"]), _n(p["cacheCreate"])) for p in t["byProject"]]
    out.append(_table(["proj", "totalTokens", "cacheRead", "cacheCreate"], rows))
    out.append(_dropped_note(t["byProjectDroppedCount"]))
    return "\n".join(out)


def _render_wall_clock(w, top_n):
    out = [f"## Wall clock (top {top_n})\n"]
    out.append(f"Total tool time: {_n(w['totalToolSec'])}s. {w['note']}\n")
    out.append("\nBy tool:\n")
    out.append(_table(["tool", "totalSec"], [(r["tool"], _n(r["totalSec"])) for r in w["byTool"]]))
    out.append("\nBy project:\n")
    out.append(_table(["proj", "totalSec"], [(r["proj"], _n(r["totalSec"])) for r in w["byProject"]]))
    out.append(_dropped_note(w["byProjectDroppedCount"]))
    return "\n".join(out)


def _render_commands(c, top_n):
    out = [f"## Commands (top {top_n})\n"]
    out.append("\nFamilies by total time:\n")
    out.append(_table(
        ["family", "bin", "count", "totalSec", "medianSec", "p95Sec", "errRate"],
        [(r["family"], r["bin"], _n(r["count"]), _n(r["totalSec"]), _n(r["medianSec"]), _n(r["p95Sec"]), _n(r["errRate"]))
         for r in c["familiesByTotalSec"]],
    ))
    out.append(_dropped_note(c["familiesByTotalSecDroppedCount"]))
    out.append("\nFamilies by count:\n")
    out.append(_table(
        ["family", "bin", "count", "totalSec"],
        [(r["family"], r["bin"], _n(r["count"]), _n(r["totalSec"])) for r in c["familiesByCount"]],
    ))
    out.append(_dropped_note(c["familiesByCountDroppedCount"]))
    out.append("\nRepeated identical commands inside one session:\n")
    out.append(_table(
        ["sid", "cmd", "occurrences", "wastedSec"],
        [(r["sid"][:8], r["cmd"][:60], _n(r["occurrences"]), _n(r["wastedSec"])) for r in c["repeatsInSession"]],
    ))
    out.append(_dropped_note(c["repeatsInSessionDroppedCount"]))
    out.append("\nFlag flailing (>=4 distinct commands of the same binary within 10 minutes):\n")
    out.append(_table(
        ["sid", "bin", "distinctCommands", "occurrences", "totalSec"],
        [(r["sid"][:8], r["bin"], _n(r["distinctCommands"]), _n(r["occurrences"]), _n(r["totalSec"])) for r in c["flagFlailing"]],
    ))
    out.append(_dropped_note(c["flagFlailingDroppedCount"]))
    return "\n".join(out)


def _render_errors(e, top_n):
    out = [f"## Errors (top {top_n})\n"]
    out.append("\nClusters:\n")
    out.append(_table(
        ["sig", "count", "sessionCount", "topTool", "sampleMsg"],
        [(r["sig"][:60], _n(r["count"]), _n(r["sessionCount"]), r.get("topTool", ""), r["sampleMsg"][:80])
         for r in e["clusters"]],
    ))
    out.append(_dropped_note(e["clustersDroppedCount"]))
    out.append("\nMissing binaries:\n")
    out.append(_table(["bin", "count", "sample"], [(r["bin"], _n(r["count"]), r["sample"][:80]) for r in e["missingBinaries"]]))
    out.append("\nPermission denied:\n")
    out.append(_table(["family", "count", "sample"], [(r["family"], _n(r["count"]), r["sample"][:80]) for r in e["permissionDenied"]]))
    out.append("\nRetry loops (same command re-run within 5 minutes of an error):\n")
    out.append(_table(["family", "count", "medianRetries"], [(r["family"], _n(r["count"]), _n(r["medianRetries"])) for r in e["retryLoops"]]))
    return "\n".join(out)


def _render_context(c, top_n):
    out = [f"## Context (top {top_n})\n"]
    out.append("\nSessions by largest single-request context:\n")
    out.append(_table(["sid", "proj", "kind", "maxCtx"], [(r["sid"][:8], r["proj"][:40], r["kind"], _n(r["maxCtx"])) for r in c["topSessions"]]))
    out.append(_dropped_note(c["topSessionsDroppedCount"]))
    out.append("\nBiggest tool outputs:\n")
    out.append(_table(["sid", "tool", "arg", "outChars"], [(r["sid"][:8], r["tool"], (r.get("arg") or "")[:50], _n(r["outChars"])) for r in c["bigOutputs"]]))
    out.append(_dropped_note(c["bigOutputsDroppedCount"]))
    out.append("\np95 output chars per tool:\n")
    out.append(_table(["tool", "p95Chars"], [(t, _n(v)) for t, v in c["bigOutputsP95PerTool"].items()]))
    out.append("\nRe-reads of the same file in one session:\n")
    out.append(_table(["sid", "arg", "occurrences", "wastedTokensEst"], [(r["sid"][:8], r["arg"][:50], _n(r["occurrences"]), _n(r["wastedTokensEst"])) for r in c["reReads"]]))
    out.append(_dropped_note(c["reReadsDroppedCount"]))
    out.append("\nCache-write ratio by kind (cacheCreate / (cacheCreate + cacheRead)):\n")
    out.append(_table(["kind", "ratio"], [(k, _n(v)) for k, v in c["cacheWriteRatio"].items()]))
    return "\n".join(out)


def _render_web(w, top_n):
    out = [f"## Web (top {top_n})\n"]
    out.append(f"Counts: {w['counts']}. Total time: {_n(w['totalSec'])}s. Total chars: {_n(w['totalChars'])}.\n")
    out.append("\nRepeated fetches/searches:\n")
    out.append(_table(["tool", "key", "count", "chars", "crossRun"], [(r["tool"], r["key"][:60], _n(r["count"]), _n(r["chars"]), r["crossRun"]) for r in w["repeats"]]))
    out.append(_dropped_note(w["repeatsDroppedCount"]))
    out.append("\nConcentration by workflow run:\n")
    out.append(_table(["run", "count"], [(r["run"], _n(r["count"])) for r in w["concentrationByRun"]]))
    out.append(_dropped_note(w["concentrationByRunDroppedCount"]))
    return "\n".join(out)


def _render_fanout(f, top_n):
    out = [f"## Fan-out (top {top_n})\n"]
    out.append("\nWorkflow runs by estimated cost:\n")
    out.append(_table(
        ["run", "agents", "tokens", "estCostUsd", "medianFirstTurnCacheCreate"],
        [(r["run"], _n(r["agents"]), _n(r["tokens"]), _usd(r["estCostUsd"]), _n(r["medianFirstTurnCacheCreate"])) for r in f["byRun"]],
    ))
    out.append(_dropped_note(f["byRunDroppedCount"]))
    out.append("\nLow-yield agents (bottom decile outTokens/cacheReadTokens):\n")
    out.append(_table(["sid", "run", "outTokens", "cacheReadTokens", "ratio"], [(r["sid"][:8], r["run"], _n(r["outTokens"]), _n(r["cacheReadTokens"]), _n(r["ratio"])) for r in f["lowYield"]]))
    out.append(_dropped_note(f["lowYieldDroppedCount"]))
    return "\n".join(out)


def _render_human(h, top_n):
    out = [f"## Human (top {top_n})\n"]
    out.append(f"Human turns: {_n(h['humanTurnCount'])}. Interrupts: {_n(h['interrupts'])}.\n")
    out.append("\nCorrection signal buckets (a turn may match more than one bucket):\n")
    for b in h["corrections"]:
        out.append(f"- {b['bucket']}: {b['count']}")
        for s in b["samples"]:
            out.append(f"  - \"{s}\"")
    out.append("\n\nRepeated near-duplicate requests across sessions:\n")
    out.append(_table(["clusterKey", "count", "sids"], [(r["clusterKey"][:60], _n(r["count"]), ", ".join(s[:8] for s in r["sids"][:5])) for r in h["repeatedRequests"]]))
    out.append(_dropped_note(h["repeatedRequestsDroppedCount"]))
    return "\n".join(out)


def _render_projects(rows, top_n):
    out = [f"## Projects (top {top_n} of {len(rows)})\n"]
    shown = rows[:top_n]
    table_rows = []
    for r in shown:
        pkg = r["onDisk"]["packageJson"]
        pm = pkg["packageManager"] if pkg else ""
        table_rows.append((
            r["proj"][:40], _n(r["sessions"]), _n(r["tokens"]), _n(r["bashSec"]),
            ", ".join(r["topCommands"]), "CLAUDE.md" if r["onDisk"]["claudeMd"] else "",
            "AGENTS.md" if r["onDisk"]["agentsMd"] else "", pm or "",
        ))
    out.append(_table(["proj", "sessions", "tokens", "bashSec", "topCommands", "claudeMd", "agentsMd", "pkgManager"], table_rows))
    if len(rows) > top_n:
        out.append(_dropped_note(len(rows) - top_n))
    return "\n".join(out)


def render_md(agg, top_n) -> str:
    """Compact markdown report body. Caller appends the work dir path as the
    final stdout line separately (see CLI contract).
    """
    parts = [
        "# harness-doctor audit\n",
        _render_window(agg["window"]),
        _render_tokens(agg["tokens"], top_n),
        _render_wall_clock(agg["wallClock"], top_n),
        _render_commands(agg["commands"], top_n),
        _render_errors(agg["errors"], top_n),
        _render_context(agg["context"], top_n),
        _render_web(agg["web"], top_n),
        _render_fanout(agg["fanout"], top_n),
        _render_human(agg["human"], top_n),
        _render_projects(agg["projects"], top_n),
    ]
    return "\n\n".join(parts)
